commec/json_io.py

```diff
@@ -0,0 +1,309 @@
+    outputs. Information is stored as a structure of dataclasses, and are 
+    converted between the dataclass / dict / json_file as required. The 
+    conversions are done dynamically, and it is recommended to only use and 
+    interact with the dataclasses only, to maintain version format, and not 
+    create erroneous outputs to the JSON which wont be read back in. This
+    ensures an expected i/o behaviour.
+
+   In this way, the JSON object serves as a common state, that can be updated
+    whilst not being temporally appended like a log file i.e. .screen file.
+'''
+
+import json
+import string
+from dataclasses import dataclass, asdict, fields, field, is_dataclass
+import os
+from typing import Dict, Type, get_origin, Any, get_args
+
+# Seperate versioning for the output JSON.
+JSON_COMMEC_FORMAT_VERSION = "1.0"
+
+@dataclass
+class MatchRange:
+    '''Container for information on how a match maps to a query, match direction
+      is +1, or positive, if the query and match are in the same direction, and 
+      -1 if they are inverse, and 0 if not set.'''
+    match_start : int = 0
+    match_end : int = 0
+    query_start : int = 0
+    query_end : int = 0
+    query_direction : int = 0
+
+@dataclass
+class BenignData:
+    '''Container to hold data related to the benign status of a taxonomy search'''
+    benign : bool = False
+    benign_match_type : str = ""
+    percent_identity_e_value : float = 0.0
+
+@dataclass
+class MatchFields:
+    '''Container to hold information for a match to the query identified in a commec screen. 
+    molecule_alphabet is expected to contain either 'aa' for aminoacids, or 'nt' for nucelotides.
+    range contains information pertaining to over what range the match occured.'''
+    molecule_alphabet : str = ""
+    description : str = ""
+    id : int = 0
+    taxon : str = ""
+    kingdom : str = ""
+    range : MatchRange = field(default_factory = MatchRange)
+    is_benign : BenignData = field(default_factory = BenignData)
+    # etc etc
+
+@dataclass
+class BioRiskData:
+    '''Container dataclass for a list of matches to biorisks 
+    identified from a commec database screen.'''
+    matches : list[MatchFields] = field(default_factory = list)
+
+@dataclass
+class TaxonomyData:
+    '''Container dataclass for a list of matches of taxonomy hits, 
+    identified from a commec database screen.'''
+    is_regulated : bool
+    matches : list[MatchFields] = field(default_factory = list)
+
+@dataclass
+class QueryData:
+    '''Container to hold data related to the Query
+      used in a commec database screen'''
+    name : str = ""
+    length : int = 0
+    sequence : str = ""
+
+@dataclass
+class CommecRunInformation:
+    '''Container dataclass to hold general run information for a commec screen '''
+    commec_version : str = "0.1.2"
+    json_output_version : str = JSON_COMMEC_FORMAT_VERSION
+    # add other settings / run parameters.
```

Comment:
Would love to get some thoughts from you on how to add versions of the databases to this, as well; not sure if we can read those from blast+ somehow? (and not sure how best to version a dataset that isn't in a standard SQL/noSQL format, e.g. the biorisk and benign dbs)

Reply:
Will incorporate this when the database handlers in python are made =)